woocommerce/woocommerce
# Changelog CLI: Add command to make entry from PR description

### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:



Closes https://github.com/woocommerce/woocommerce/issues/38125

Add a CLI utility to add a changelog entry from details supplied in a PR description.

Note on validation: I'm not doing any validation of the input. the script will fail at the Jetpack Changelogger creation step if input not valid.

Note on changelog of dependencies: In a follow up I'd like to address changelogs being added to dependencies based on dependency graphs. If a dependency gets updated, it will get an entry and so will all projects consuming it.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Prerequisites
1. You must have a `GITHUB_TOKEN` environment variable available in your PATH. This needs to be set to a [personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) with repo permissions enabled.

#### Setup your fork
1. Fork this repo.
2. Push this branch to your fork (`add/changelog-determine-projects`).
3. Create a branch off this one called `test/add/changelog-determine-projects`.
4. In the new branch, create a few changes to the codebase. Be sure to include some changes to projects with changelogger like `woocommerce` and some to projects that do not, like any of our `tools/`.
5. Push `test/add/changelog-determine-projects` to your fork using `--no-verify` to skip changelog checks.
6. Create a pull request on your fork to merge `test/add/changelog-determine-projects` into `add/changelog-determine-projects`.
7. Note the pull request number

#### Test the command
1. Switch back to this PR's branch, `add/changelog-determine-projects` and run the command. The PR doesn't have the corresponding description check, so nothing will happen. Confirm the message makes sense.
```
pnpm run utils changefile <prNumber> -o <forkOwner>
```
2. Add this template to the bottom of the open PR description and fill in the relevant details. Confirm the data entry aspect is ok.
```
### Changelog entry




<details>

#### Significance


#### Type


#### Message


#### Comment


</details>
```
3. Now run the command again. This time it should succeed in creating changelog entries and adding directly to the pull request's branch. Confirm changelogs were added only to public projects that have Jetpack Changelogger.
```
pnpm run utils changefile <prNumber> -o <forkOwner>
```
4. In the `test/add/changelog-determine-projects` branch undo the change to `woocommerce`, if applicable. If not, undo any of the other changes. Push changes to your fork and re run the command from `add/changelog-determine-projects` to see the changelog for the project that no longer has changes go away.
```
pnpm run utils changefile <prNumber> -o <forkOwner>
```
6. Finally make a change in `test/add/changelog-determine-projects` to woocommerce-admin. Push and rerun the command. A changelog entry to `woocommerce` should be added.
